### PR TITLE
[Test][Reduction] add var/std/var_mean conformance across dim x correction

### LINKED ISSUE #1397: [FEAT][REDUCTION] implement dim=none for variance reductions (var/std/var_mean)
### Symptom / Motivation

`tileops/manifest/reduction.yaml` declares `VarFwdOp`, `StdFwdOp`, `VarMeanFwdOp` with PyTorch-compatible `correction` (or `unbiased` legacy alias) parameter; `VarMeanFwdOp` returns a tuple `(var, mean)`. The op classes (`tileops/ops/reduction/reduce.py`) ship as `status: spec-only` because the multi-dim / `dim=None` path plus the Bessel correction (`correction != 0`) edge cases are not yet covered.

W2-E-flip (#1199) leftover scope, second of three impl issues. Tracker: #1142.

### Root Cause Analysis

Variance reductions need numerically-stable two-pass or Welford-style accumulation; `VarMean` further requires emitting two outputs from a single kernel call (or two coordinated kernel calls). The `correction` parameter shifts the divisor (`N - correction` instead of `N`), which the current implementation likely hardcodes as `N - 1` (unbiased) or `N` (biased) but not parametrically.

### Related Files

- `tileops/ops/reduction/reduce.py` — `VarFwdOp`, `StdFwdOp`, `VarMeanFwdOp` classes
- `tileops/kernels/reduction/` — kernel implementations
- `tileops/manifest/reduction.yaml` — spec entries
- `tests/ops/test_reduction.py` — spec tests

## Goal

`VarFwdOp`, `StdFwdOp`, `VarMeanFwdOp` pass spec-conformance tests for every workload × dim × correction combination in the manifest. `VarMeanFwdOp` returns a tuple `(var, mean)` matching `torch.var_mean`. Numerics match `torch.var / torch.std / torch.var_mean` within tolerances. Status flip is the sibling issue's job.

## Plan

1. Run `/audit-family REDUCTION VarFwdOp StdFwdOp VarMeanFwdOp` and confirm the gap classification.
2. Run `/align-family REDUCTION VarFwdOp StdFwdOp VarMeanFwdOp` — internally delegates to `/align-op` per op.
3. For `VarMeanFwdOp`: confirm the op layer correctly returns a 2-tuple `(var, mean)` and that the kernel either fuses the two reductions or coordinates two passes without re-reading the input twice from HBM.
4. For `VarFwdOp` / `StdFwdOp`: ensure `correction` parameter is honored (defaults to 1 per PyTorch).
5. Confirm `pytest tests/ops/test_reduction.py -m smoke` passes for the 3 ops.
6. Confirm `python scripts/validate_manifest.py` exits 0.

## Constraints

- MUST NOT change manifest signature, shape_rules, dtype_combos, roofline, or workloads. Code-only PR.
- MUST NOT touch the other 7 spec-only reduction ops (separate impl issues).
- MUST match `torch.var / torch.std / torch.var_mean` numerics including the `correction` parameter (defaults to 1).
- `VarMeanFwdOp` MUST return a tuple in the same field order as PyTorch (`(var, mean)`).
- Single PR. Single nightshift.

## Acceptance Criteria

- [ ] AC-1: Modified files pass unit tests (`pytest tests/ops/test_reduction.py -m smoke` covers the 3 ops).
- [ ] AC-2: `python scripts/validate_manifest.py` exits 0; no new warnings on the 3 ops.
- [ ] AC-3: `pytest tests/test_validate_manifest.py` passes.
- [ ] AC-4: Spec-conformance tests exist for each of the 3 ops covering `dim=None`, `dim=int`, `dim=tuple`, `correction in {0, 1, 2}`, `keepdim` both.
- [ ] AC-5: `VarMeanFwdOp` test asserts the returned tuple shape/dtype matches `torch.var_mean(...)`.
- [ ] AC-6: Op classes still at `status: spec-only` (status flip is the sibling PR).

### PR DESCRIPTION
**Pull request by lcy-seso**
Closes #1397

## Summary

- Add spec-conformance tests for `VarFwdOp`, `StdFwdOp`, `VarMeanFwdOp` covering the full `dim` x `correction` x `keepdim` x `dtype` matrix.
- New file: `tests/ops/test_reduce_variance_conformance.py` (171 parametrized cases, all `smoke`-tagged).
- `VarMeanFwdOp` test asserts the returned tuple shape and dtype match `torch.var_mean`.
- Manifest entries remain at `status: spec-only`; status flip is the sibling implementation PR.

## Test plan

- [x] AC-1: `pytest tests/ops/test_reduce_variance_conformance.py -m smoke` — 171 passed
- [x] AC-2: `python scripts/validate_manifest.py` — exit 0, no new warnings on Var/Std/VarMean
- [x] AC-3: `pytest tests/test_validate_manifest.py` — 218 passed
- [x] AC-4: parametrize matrix covers `dim in {None, int, tuple}` x `correction in {0, 1, 2}` x `keepdim in {False, True}` for each of the 3 ops
- [x] AC-5: tuple/shape/dtype assertions vs `torch.var_mean` reference
- [x] AC-6: manifest entries remain `status: spec-only` (zero manifest changes)

### CODE REVIEW

File: tests/ops/test_reduce_variance_conformance.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+def _ref_var_mean(
+    x: torch.Tensor, dim, keepdim: bool, correction: int,
+) -> tuple[torch.Tensor, torch.Tensor]:
+    dims = _normalize_dims(dim, x.ndim)
+    var = torch.var(x.float(), dim=dims, keepdim=keepdim, correction=correction).to(x.dtype)
+    mean = torch.mean(x.float(), dim=dims, keepdim=keepdim).to(x.dtype)
+    return var, mean
+
+
+@pytest.mark.smoke
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_reduce_variance_conformance.py:65]: Out of scope. The whole TileOPs CUDA test suite (`tests/ops/test_reduce_*`, `test_norm_*`, attention, etc.) instantiates ops with `device='cuda'` without a skipif guard — gating CI is via the `gpu-smoke` job and the smoke marker, not per-file skipifs. Adding one here would diverge from the project convention without solving anything (GPU-less CI doesn't run these files anyway).

File: tests/ops/test_reduce_variance_conformance.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        f"mean shape {mean_y.shape} vs ref {ref_mean.shape}"
+    )
+    assert var_y.dtype == ref_var.dtype, (
+        f"var dtype {var_y.dtype} vs ref {ref_var.dtype}"
+    )
+    assert mean_y.dtype == ref_mean.dtype, (
+        f"mean dtype {mean_y.dtype} vs ref {ref_mean.dtype}"
+    )
+    torch.testing.assert_close(var_y, ref_var, **_tol(dtype))
+    torch.testing.assert_close(mean_y, ref_mean, **_tol(dtype))
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_reduce_variance_conformance.py:176]: Out of scope. `test_var_*`/`test_std_*`/`test_var_mean_*` look similar but call different ops (`VarFwdOp` returns one tensor, `VarMeanFwdOp` returns a 2-tuple, `StdFwdOp` is its own kernel binding) and use distinct ref helpers. The 0-D test in lines 184-200 already factors what was uniform; collapsing the rest would either lose readable assertion failures or reintroduce branching on op_cls inside the test body. Tracking deferred.

File: tests/ops/test_reduce_variance_conformance.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    dtype = torch.float16
+    x = torch.randn(*_UNALIGNED_SHAPE, dtype=dtype, device="cuda")
+    op = VarMeanFwdOp(dtype=dtype, dim=dim, correction=1, keepdim=False)
+    var_y, mean_y = op(x)
+    ref_var, ref_mean = _ref_var_mean(x, dim, False, 1)
+    assert var_y.shape == ref_var.shape
+    assert mean_y.shape == ref_mean.shape
+    torch.testing.assert_close(var_y, ref_var, **_tol(dtype))
+    torch.testing.assert_close(mean_y, ref_mean, **_tol(dtype))
+
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_reduce_variance_conformance.py:228]: Adopted. Added `test_std_unaligned_innermost` (3 cells: dim ∈ {int, tuple, None}) on `(4, 8, 255)`, mirroring `test_var_unaligned_innermost`. See 9a63bb3. 3/3 pass locally.

File: tests/ops/test_reduce_variance_conformance.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+        return {"atol": 1e-4, "rtol": 1e-4}
+    return {"atol": 1e-3, "rtol": 1e-3}
+
+
+def _normalize_dims(dim, ndim: int):
+    if dim is None:
+        return list(range(ndim))
+    if isinstance(dim, tuple):
+        return list(dim)
+    return dim
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_reduce_variance_conformance.py]: Adopted. Removed `_normalize_dims` and pass `dim` directly to `torch.var` / `torch.std` / `torch.var_mean`. See 8b4e096.

File: tests/ops/test_reduce_variance_conformance.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    return torch.std(x.float(), dim=dims, keepdim=keepdim, correction=correction).to(x.dtype)
+
+
+def _ref_var_mean(
+    x: torch.Tensor, dim, keepdim: bool, correction: int,
+) -> tuple[torch.Tensor, torch.Tensor]:
+    dims = _normalize_dims(dim, x.ndim)
+    var = torch.var(x.float(), dim=dims, keepdim=keepdim, correction=correction).to(x.dtype)
+    mean = torch.mean(x.float(), dim=dims, keepdim=keepdim).to(x.dtype)
+    return var, mean
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_reduce_variance_conformance.py]: Adopted. `_ref_var_mean` now calls `torch.var_mean` directly. See 8b4e096.

File: tests/ops/test_reduce_variance_conformance.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    if dim is None:
+        return list(range(ndim))
+    if isinstance(dim, tuple):
+        return list(dim)
+    return dim
+
+
+def _ref_var(x: torch.Tensor, dim, keepdim: bool, correction: int) -> torch.Tensor:
+    dims = _normalize_dims(dim, x.ndim)
+    return torch.var(x.float(), dim=dims, keepdim=keepdim, correction=correction).to(x.dtype)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_reduce_variance_conformance.py]: Adopted. `_ref_var` simplified to a direct `torch.var` call. See 8b4e096.

File: tests/ops/test_reduce_variance_conformance.py [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+
+
+def _ref_std(x: torch.Tensor, dim, keepdim: bool, correction: int) -> torch.Tensor:
+    dims = _normalize_dims(dim, x.ndim)
+    return torch.std(x.float(), dim=dims, keepdim=keepdim, correction=correction).to(x.dtype)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_reduce_variance_conformance.py]: Adopted. `_ref_std` simplified to a direct `torch.std` call. See 8b4e096.

File: tests/ops/test_reduce_variance_conformance.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+_UNALIGNED_SHAPE = (4, 8, 255)
+
+
+def _tol(dtype: torch.dtype) -> dict:
+    # Welford accumulators run in fp32 and only narrow at the boundary, so
+    # half-precision tolerances stay near the unit in the last place of the
+    # storage dtype rather than the looser 1e-2 default.
+    if dtype == torch.float32:
+        return {"atol": 1e-4, "rtol": 1e-4}
+    return {"atol": 1e-3, "rtol": 1e-3}
```
Comment (Ibuki-wind) [inline tests/ops/test_reduce_variance_conformance.py]: `_tol()` returns `1e-3` for `torch.bfloat16`, which is stricter than the repository’s reduction-test guidance and makes these smoke cases brittle on real GPUs -> use the BF16 tolerance already used by the neighboring Welford tests (for example `atol=rtol=1.6e-2`) or a shared helper.

File: tests/ops/test_reduce_variance_conformance.py [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+        pytest.param(StdFwdOp, _ref_std, id="StdFwdOp"),
+    ],
+)
+def test_dim_none_keepdim_false_returns_0d(op_cls, ref_fn) -> None:
+    """``dim=None, keepdim=False`` must return a 0-D tensor matching PyTorch."""
+    torch.manual_seed(0)
+    x = torch.randn(*_SHAPE, dtype=torch.float32, device="cuda")
+    op = op_cls(dtype=torch.float32, dim=None, correction=1, keepdim=False)
+    y = op(x)
+    ref = ref_fn(x, None, False, 1)
```
Comment (Ibuki-wind) [inline tests/ops/test_reduce_variance_conformance.py]: `test_dim_none_keepdim_false_returns_0d` and `test_var_mean_dim_none_keepdim_false_returns_0d` duplicate the `dim=None, keepdim=False` branch already asserted by the matrix tests above, so they only add nodes without new coverage -> fold the 0-D assertion into the existing parametrized cases or drop these standalone tests.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
### Overall
Two blocking test-hygiene issues need cleanup before merge.

File: tests/ops/test_reduce_variance_conformance.py (hunk 8 above)
Comment (lcy-seso) [reply tests/ops/test_reduce_variance_conformance.py]: Adopted. Reverted bf16/fp16 tolerance to 1e-2 to match the shared convention used by test_reduce_dim_none.py / test_reduce_multidim.py / test_reduce_arithmetic_conformance.py. See be13fab.

File: tests/ops/test_reduce_variance_conformance.py (hunk 9 above)
Comment (lcy-seso) [reply tests/ops/test_reduce_variance_conformance.py]: Adopted. Dropped both `test_dim_none_keepdim_false_returns_0d` and `test_var_mean_dim_none_keepdim_false_returns_0d` — the dim=None / keepdim=False cells are already covered by the matrix tests above. See be13fab.

**Review by Ibuki-wind — APPROVED** [review]
Clean — no issues.
